File: lc-cpp/problems/1/1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

#include "1.h"

using namespace std;

vector<int> twoSum(vector<int> &nums, int target)
{
    vector<int> result;

    unordered_map<int, int> m;
    int i = 0;
    for (int &n : nums)
    {
        size_t v = m.count(n);
        if (v)
        {
            return {m[n], i};
        }
        m[target - n] = i;
        ++i;
    }

    return result;
}

int main()
{
    if (!test_1())
    {
        cout << "test 1 failed\n";
    }
    else
    {
        cout << "test 1 passed\n";
    }

    if (!test_2())
    {
        cout << "test 2 failed\n";
    }
    else
    {
        cout << "test 2 passed\n";
    }

    if (!test_3())
    {
        cout << "test 3 failed\n";
    }
    else
    {
        cout << "test 3 passed\n";
    }

    return 0;
}

bool test_1()
{
    vector<int> nums = {2,
                        7,
                        11,
                        15};
    auto res = twoSum(nums, 9);
    if (res.size() != 2)
        return false;
    if (res[0] != 0 || res[1] != 1)
        return false;

    return true;
}

bool test_2()
{
    vector<int> nums = {3,
                        2,
                        4};
    auto res = twoSum(nums, 6);
    if (res.size() != 2)
        return false;
    if (res[0] != 1 || res[1] != 2)
        return false;

    return true;
}

bool test_3()
{
    vector<int> nums = {3,
                        3};
    auto res = twoSum(nums, 6);
    if (res.size() != 2)
        return false;
    if (res[0] != 0 || res[1] != 1)
        return false;

    return true;
}

File: lc-lua/problems/1/1.lua
function twoSum(nums, target)
    local m = {}
    for i, n in pairs(nums) do
        local v = m[n]
        if v then
            return {v, i}
        end
        m[target - n] = i
    end
    return nil
end

-- I really like that everything is a table in this language
-- but, ~= and 1 index are weird decisions

local test_1_result = twoSum({2, 7, 11, 15}, 9)
if (test_1_result[1] ~= 1 or test_1_result[2] ~= 2) then
    print("test 1 failed")
else 
    print("test 1 passed")
end

local test_2_result = twoSum({3, 2, 4}, 6)
if (test_2_result[1] ~= 2 or test_2_result[2] ~= 3) then
    print("test 2 failed")
else 
    print("test 2 passed")
end

local test_3_result = twoSum({3, 3}, 6)
if (test_3_result[1] ~= 1 or test_3_result[2] ~= 2) then
    print("test 3 failed")
else 
    print("test 3 passed")
end